xls/modules/rle/rle_enc_adv.md

```diff
@@ -0,0 +1,85 @@
+# Multisymbol Run-length Encoder
+
+The encoder uses Run Length Encoding (RLE) to compress the input stream of
+repeating symbols to the output stream that contains the symbols and
+the number of its consecutive occurrences in the input stream.
+
+Overall, we can break down the data processing into four stages: reduction, alignment, compression, and output generation. The division of responsibility allowed the specialized blocks to efficiently process data and made it possible test each functionality separately.
+
+The first block is responsible for taking the input and reducing it to pairs of symbols and the number of its occurrences, as in the [basic RLE implementation](https://github.com/google/xls/pull/974). The second element of the encoder shifts the previously emitted pairs and adjusts them for further processing. Both of these elements have an empty state. The next block takes the prepared data and combines it with the information about previously processed symbols. The last element is responsible for adjusting the width of the output data to the receiver interface.
+
+
+Input width is defined using the `INPUT_WIDTH` parameter and output width is defined with the `OUTPUT_WIDTH` parameter.
+Both the input and the output channels use additional `last` flag
+that indicates whether the packet ends the transmission. After sending
+the last packet the encoder dumps all the data to the output stream.
+
+## Encoder processing pipeline detailed breakdown.
+
+### Initial conditions
+- input width is 4 symbols wide,
+- output width is 2 pairs wide,
+- symbol counter is 2 bits wide.
+
+### Process
+1. Reduce step - this process takes incoming symbols and symbol_valid
+and reduces them into symbol count pairs. This step is stateless.
```

Comment:
couldn't stateless proc be in theory implemented as `fn`?

Reply:
All state less processes are effectively a `fn`s, but as far as I know you can't put `recv/send` operations into `fn`s.
Using `fn ` would potentially reduce fmax, as first proc would have to do all the processing before it would be able to update it's state, or we would reduce pipeline throughput when using II>1 to improve fmax.